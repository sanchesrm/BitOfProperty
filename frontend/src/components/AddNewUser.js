import React, { Component, Fragment } from "react";
import { Form, Button, Col, Row, Input, message } from "antd";
import { createUser, fetchUsers } from "../actions/userActions";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";

export class AddNewUser extends Component {
  constructor() {
    super();

    this.state = {
      addingUserLoading: false
    };
  }

  handleSubmit = e => {
    e.preventDefault();
    const { form, closeDrawer } = this.props;
    form.validateFields((err, values) => {
      if (!err) {
        this.setState({
          addingUserLoading: true
        });

        this.props
          .createUser(values.email)
          .then(() => {
            form.resetFields();
            closeDrawer();
          })
          .then(() => {
            this.props.fetchUsers();
          })
          .catch(() => {
            message.error("Something went wrong!", 2.5);
          })
          .finally(() => {
            this.setState({
              addingUserLoading: false
            });
          });
      }
    });
  };

  render() {
    const {
      closeDrawer,
      form: { getFieldDecorator }
    } = this.props;

    return (
      <Fragment>
        <Form onSubmit={this.handleSubmit} layout="vertical" hideRequiredMark>
          <Row>
            <Col>
              <Form.Item label="Username">
                {getFieldDecorator("email", {
                  rules: [
                    {
                      type: "email",
                      message: "Username should be an e-mail!"
                    },
                    {
                      required: true,
                      message: "Please enter username"
                    }
                  ]
                })(<Input placeholder="Please enter username" />)}
              </Form.Item>
            </Col>
          </Row>
          <Form.Item>
            <Button onClick={closeDrawer} style={{ marginRight: 8 }}>
              Cancel
            </Button>
            <Button
              type="primary"
              htmlType="submit"
              loading={this.state.addingUserLoading}
            >
              Create User
            </Button>
          </Form.Item>
        </Form>
      </Fragment>
    );
  }
}

export const AddNewUserForm = Form.create()(AddNewUser);

export const mapStateToProps = ({ users }) => {
  return { users };
};

export const mapDispatchToProps = dispatch => {
  return bindActionCreators({ createUser, fetchUsers }, dispatch);
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AddNewUserForm);
