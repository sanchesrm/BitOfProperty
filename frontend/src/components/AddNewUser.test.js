import Adapter from "enzyme-adapter-react-16";
import React from "react";
import toJson from "enzyme-to-json";
import { configure, mount } from "enzyme";
import { AddNewUserForm, AddNewUser, mapStateToProps } from "./AddNewUser";
import { Form } from "antd";

describe("AddNewUser Component", () => {
  let props;
  beforeEach(() => {
    configure({ adapter: new Adapter() });
    props = {
      createUser: jest.fn(() => Promise.resolve()),
      fetchUsers: jest.fn(() => Promise.resolve()),
      closeDrawer: jest.fn(),
      history: {
        push: jest.fn()
      }
    };
  });

  test("render correctly", () => {
    expect.assertions(1);

    const EnhancedForm = Form.create()(AddNewUserForm);
    const wrapper = mount(<EnhancedForm {...props} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test("should validate required fields", () => {
    expect.assertions(1);

    const EnhancedForm = Form.create()(AddNewUserForm);
    const wrapper = mount(<EnhancedForm {...props} />);

    const createUserMethod = jest.spyOn(props, "createUser");
    const AddNewUserWrapped = wrapper.wrap(wrapper.find(AddNewUser));

    AddNewUserWrapped.instance().handleSubmit({ preventDefault: jest.fn() });
    expect(createUserMethod).not.toHaveBeenCalled();
  });

  test("should validate required fields and pass", () => {
    expect.assertions(3);

    const EnhancedForm = Form.create()(AddNewUserForm);
    const wrapper = mount(<EnhancedForm {...props} />);
    const AddNewUserWrapped = wrapper.wrap(wrapper.find(AddNewUser));

    const createUserMethod = jest.spyOn(props, "createUser");
    const fetchUsersMethod = jest.spyOn(props, "fetchUsers");
    const closeDrawerMethod = jest.spyOn(props, "closeDrawer");

    AddNewUserWrapped.find("input")
      .get(0)
      .props.onChange({
        target: { value: "bitofproperty@rocks.com" }
      });

    AddNewUserWrapped.instance().handleSubmit({ preventDefault: jest.fn() });
    expect(createUserMethod).toHaveBeenCalled();
    props
      .createUser()
      .then(() => {
        expect(closeDrawerMethod).toHaveBeenCalled();
      })
      .then(() => {
        expect(fetchUsersMethod).toHaveBeenCalled();
      });
  });

  test("should validate required fields and fail to create", () => {
    expect.assertions(2);

    props = {
      ...props,
      createUser: jest.fn(() => Promise.reject())
    };

    const EnhancedForm = Form.create()(AddNewUserForm);
    const wrapper = mount(<EnhancedForm {...props} />);
    const AddNewUserWrapped = wrapper.wrap(wrapper.find(AddNewUser));

    const createUserMethod = jest.spyOn(props, "createUser");
    const closeDrawerMethod = jest.spyOn(props, "closeDrawer");

    AddNewUserWrapped.find("input")
      .get(0)
      .props.onChange({
        target: { value: "bitofproperty@rocks.com" }
      });

    AddNewUserWrapped.instance().handleSubmit({ preventDefault: jest.fn() });
    expect(createUserMethod).toHaveBeenCalled();
    props.createUser().catch(() => {
      expect(closeDrawerMethod).not.toHaveBeenCalled();
    });
  });

  test("test mapStateToProps", () => {
    expect.assertions(1);

    const mapStateToPropsShallow = mapStateToProps({
      users: []
    });

    expect(mapStateToPropsShallow.users).toEqual([]);
  });
});
