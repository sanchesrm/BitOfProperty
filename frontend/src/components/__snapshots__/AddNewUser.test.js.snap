// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNewUser Component render correctly 1`] = `
<Form(Form(AddNewUser))
  closeDrawer={[MockFunction]}
  createUser={[MockFunction]}
  fetchUsers={[MockFunction]}
  history={
    Object {
      "push": [MockFunction],
    }
  }
>
  <Form(AddNewUser)
    closeDrawer={[MockFunction]}
    createUser={[MockFunction]}
    fetchUsers={[MockFunction]}
    form={
      Object {
        "getFieldDecorator": [Function],
        "getFieldError": [Function],
        "getFieldInstance": [Function],
        "getFieldProps": [Function],
        "getFieldValue": [Function],
        "getFieldsError": [Function],
        "getFieldsValue": [Function],
        "isFieldTouched": [Function],
        "isFieldValidating": [Function],
        "isFieldsTouched": [Function],
        "isFieldsValidating": [Function],
        "isSubmitting": [Function],
        "resetFields": [Function],
        "setFields": [Function],
        "setFieldsInitialValue": [Function],
        "setFieldsValue": [Function],
        "submit": [Function],
        "validateFields": [Function],
        "validateFieldsAndScroll": [Function],
      }
    }
    history={
      Object {
        "push": [MockFunction],
      }
    }
  >
    <AddNewUser
      closeDrawer={[MockFunction]}
      createUser={[MockFunction]}
      fetchUsers={[MockFunction]}
      form={
        Object {
          "getFieldDecorator": [Function],
          "getFieldError": [Function],
          "getFieldInstance": [Function],
          "getFieldProps": [Function],
          "getFieldValue": [Function],
          "getFieldsError": [Function],
          "getFieldsValue": [Function],
          "isFieldTouched": [Function],
          "isFieldValidating": [Function],
          "isFieldsTouched": [Function],
          "isFieldsValidating": [Function],
          "isSubmitting": [Function],
          "resetFields": [Function],
          "setFields": [Function],
          "setFieldsInitialValue": [Function],
          "setFieldsValue": [Function],
          "submit": [Function],
          "validateFields": [Function],
          "validateFieldsAndScroll": [Function],
        }
      }
      history={
        Object {
          "push": [MockFunction],
        }
      }
    >
      <Form
        hideRequiredMark={true}
        layout="vertical"
        onSubmit={[Function]}
        prefixCls="ant-form"
      >
        <form
          className="ant-form ant-form-vertical ant-form-hide-required-mark"
          onSubmit={[Function]}
        >
          <Row
            gutter={0}
          >
            <div
              className="ant-row"
            >
              <Col>
                <div
                  className=""
                >
                  <FormItem
                    colon={true}
                    hasFeedback={false}
                    label="Username"
                    prefixCls="ant-form"
                  >
                    <Row
                      className="ant-form-item"
                      gutter={0}
                    >
                      <div
                        className="ant-row ant-form-item"
                      >
                        <Col
                          className="ant-form-item-label"
                          key="label"
                        >
                          <div
                            className="ant-form-item-label"
                          >
                            <label
                              className="ant-form-item-required"
                              htmlFor="email"
                              onClick={[Function]}
                              title="Username"
                            >
                              Username
                            </label>
                          </div>
                        </Col>
                        <Col
                          className="ant-form-item-control-wrapper"
                          key="wrapper"
                        >
                          <div
                            className="ant-form-item-control-wrapper"
                          >
                            <div
                              className="ant-form-item-control"
                            >
                              <span
                                className="ant-form-item-children"
                              >
                                <Input
                                  data-__field={
                                    Object {
                                      "name": "email",
                                    }
                                  }
                                  data-__meta={
                                    Object {
                                      "name": "email",
                                      "originalProps": Object {
                                        "disabled": false,
                                        "placeholder": "Please enter username",
                                        "prefixCls": "ant-input",
                                        "type": "text",
                                      },
                                      "ref": null,
                                      "rules": Array [
                                        Object {
                                          "message": "Username should be an e-mail!",
                                          "type": "email",
                                        },
                                        Object {
                                          "message": "Please enter username",
                                          "required": true,
                                        },
                                      ],
                                      "trigger": "onChange",
                                      "validate": Array [
                                        Object {
                                          "rules": Array [
                                            Object {
                                              "message": "Username should be an e-mail!",
                                              "type": "email",
                                            },
                                            Object {
                                              "message": "Please enter username",
                                              "required": true,
                                            },
                                          ],
                                          "trigger": Array [
                                            "onChange",
                                          ],
                                        },
                                      ],
                                      "valuePropName": "value",
                                    }
                                  }
                                  disabled={false}
                                  id="email"
                                  onChange={[Function]}
                                  placeholder="Please enter username"
                                  prefixCls="ant-input"
                                  type="text"
                                >
                                  <input
                                    className="ant-input"
                                    data-__field={
                                      Object {
                                        "name": "email",
                                      }
                                    }
                                    data-__meta={
                                      Object {
                                        "name": "email",
                                        "originalProps": Object {
                                          "disabled": false,
                                          "placeholder": "Please enter username",
                                          "prefixCls": "ant-input",
                                          "type": "text",
                                        },
                                        "ref": null,
                                        "rules": Array [
                                          Object {
                                            "message": "Username should be an e-mail!",
                                            "type": "email",
                                          },
                                          Object {
                                            "message": "Please enter username",
                                            "required": true,
                                          },
                                        ],
                                        "trigger": "onChange",
                                        "validate": Array [
                                          Object {
                                            "rules": Array [
                                              Object {
                                                "message": "Username should be an e-mail!",
                                                "type": "email",
                                              },
                                              Object {
                                                "message": "Please enter username",
                                                "required": true,
                                              },
                                            ],
                                            "trigger": Array [
                                              "onChange",
                                            ],
                                          },
                                        ],
                                        "valuePropName": "value",
                                      }
                                    }
                                    disabled={false}
                                    id="email"
                                    onChange={[Function]}
                                    onKeyDown={[Function]}
                                    placeholder="Please enter username"
                                    type="text"
                                    value=""
                                  />
                                </Input>
                              </span>
                              <Animate
                                animation={Object {}}
                                component=""
                                componentProps={Object {}}
                                key="help"
                                onAppear={[Function]}
                                onEnd={[Function]}
                                onEnter={[Function]}
                                onLeave={[Function]}
                                transitionAppear={true}
                                transitionEnter={true}
                                transitionLeave={true}
                                transitionName="show-help"
                              />
                            </div>
                          </div>
                        </Col>
                      </div>
                    </Row>
                  </FormItem>
                </div>
              </Col>
            </div>
          </Row>
          <FormItem
            colon={true}
            hasFeedback={false}
            prefixCls="ant-form"
          >
            <Row
              className="ant-form-item"
              gutter={0}
            >
              <div
                className="ant-row ant-form-item"
              >
                <Col
                  className="ant-form-item-control-wrapper"
                  key="wrapper"
                >
                  <div
                    className="ant-form-item-control-wrapper"
                  >
                    <div
                      className="ant-form-item-control"
                    >
                      <span
                        className="ant-form-item-children"
                      >
                        <Button
                          block={false}
                          ghost={false}
                          loading={false}
                          onClick={[MockFunction]}
                          prefixCls="ant-btn"
                          style={
                            Object {
                              "marginRight": 8,
                            }
                          }
                        >
                          <Wave>
                            <button
                              className="ant-btn"
                              onClick={[Function]}
                              style={
                                Object {
                                  "marginRight": 8,
                                }
                              }
                              type="button"
                            >
                              <span
                                key=".0"
                              >
                                Cancel
                              </span>
                            </button>
                          </Wave>
                        </Button>
                        <Button
                          block={false}
                          ghost={false}
                          htmlType="submit"
                          loading={false}
                          prefixCls="ant-btn"
                          type="primary"
                        >
                          <Wave>
                            <button
                              className="ant-btn ant-btn-primary"
                              onClick={[Function]}
                              type="submit"
                            >
                              <span
                                key=".0"
                              >
                                Create User
                              </span>
                            </button>
                          </Wave>
                        </Button>
                      </span>
                      <Animate
                        animation={Object {}}
                        component=""
                        componentProps={Object {}}
                        key="help"
                        onAppear={[Function]}
                        onEnd={[Function]}
                        onEnter={[Function]}
                        onLeave={[Function]}
                        transitionAppear={true}
                        transitionEnter={true}
                        transitionLeave={true}
                        transitionName="show-help"
                      />
                    </div>
                  </div>
                </Col>
              </div>
            </Row>
          </FormItem>
        </form>
      </Form>
    </AddNewUser>
  </Form(AddNewUser)>
</Form(Form(AddNewUser))>
`;
